solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandlerFactory.java

```diff
@@ -0,0 +1,180 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.component;
+
+import com.codahale.metrics.Counter;
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.ExecutorService;
+import org.apache.solr.common.util.ExecutorUtil;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.apache.solr.core.PluginInfo;
+import org.apache.solr.core.SolrInfoBean;
+import org.apache.solr.metrics.SolrMetricManager;
+import org.apache.solr.metrics.SolrMetricsContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TimeLimitingHttpShardHandlerFactory extends HttpShardHandlerFactory {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private long slowNodeTimeout;
+  private boolean dryRun;
+  private boolean initialized;
+
+  private static final String TIMEOUT_CONFIG_KEY =
+      "slowNodeTimeout"; // config key for timeout in millisec
+  private static final String LATENCY_DROP_RATIO_THRESHOLD_CONFIG_KEY = "latencyDropRatioThreshold";
+  private static final String MAX_SLOW_RESPONSE_PERCENTAGE_CONFIG_KEY = "maxSlowResponsePercentage";
+  private static final String MIN_SHARD_COUNT_PER_REQUEST_CONFIG_KEY = "minShardCountPerRequest";
+  private static final String SLOW_LATENCY_THRESHOLD_CONFIG_KEY = "slowLatencyThreshold";
+  private static final String SLOW_NODE_TTL_CONFIG_KEY = "slowNodeTtl";
+  private static final String DRY_RUN_CONFIG_KEY = "dryRun";
+
+  private SlowNodeDetector slowNodeDetector;
+  private SolrMetricsContext solrMetricsContext;
+  Counter cancelledSlowNodeRequests;
+  Counter cancelledDryRunSlowNodeRequests;
+  private ExecutorService executorService;
+
+  /**
+   * Get {@link ShardHandler} that times out on slow shards.
+   *
+   * <p>Take note the returned ShardHandler is expected to handle ShardRequests spawned by a single
+   * top level query.
+   *
+   * <p>Each ShardRequest are expected to be submitted to all the relevant cores in relatively close
+   * timeframes (ie most requests are submitted before the first response comes back) Otherwise, it
+   * might not detect slow nodes nor timeout slow requests as expected.
+   */
+  @Override
+  public ShardHandler getShardHandler() {
+    if (!initialized) {
+      throw new RuntimeException(
+          TimeLimitingHttpShardHandlerFactory.class.getSimpleName()
+              + " is not initialized, run init() first or check if there are any exceptions during init()");
+    }
+    return new TimeLimitingHttpShardHandler(
+        this,
+        slowNodeTimeout,
+        dryRun,
+        slowNodeDetector,
+        (timedOutTasks) -> {
+          Counter counter = dryRun ? cancelledDryRunSlowNodeRequests : cancelledSlowNodeRequests;
+          counter.inc(timedOutTasks.size());
+        },
```

Comment:
one other thought: it seems a little odd to me that we're passing `Set<Future<?>>` here, only to unconditionally invoke `timedOutTasks.size()`. Unless we anticipate using the set in some other way, could we just make this an `IntConsumer`? (and possibly also clarifying that this callback is strictly about metrics (since it seems that's the case with the current code).

Not a big deal either way.